File: TopCoder/python/arr/RemoveDuplicatesFromSortedList.py

from math import inf

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def deleteDuplicates(self, head: ListNode) -> ListNode:
        if not head:
            return None
        if not head.next:
            return head
        fakeNode = ListNode(0)
        prev_val = inf

        current_temp = fakeNode

        while head is not None:
            if head.next is not None:
                if head.val == head.next.val or prev_val == head.val:
                    prev_val = head.val
                else:
                    current_temp.next = head
                    prev_val = head.val
                    current_temp = current_temp.next
            else:
                # last node
                if prev_val == head.val:
                    current_temp.next = None
                else:
                    current_temp.next = head
            prev_val = head.val
            head = head.next

        return fakeNode.next

if __name__ == "__main__":

    node = ListNode(1)
    node.next = ListNode(1)
    node.next.next = ListNode(2)
    #node.next.next.next = ListNode(2)
    #node.next.next.next.next = ListNode(3)

    sol = Solution()
    reesult = sol.deleteDuplicates(node)
    while reesult:
        print(reesult.val)
        reesult = reesult.next

File: TopCoder/src/main/java/com/leetcode/linkedlist/RemoveDuplicatesFromSortedList.java
package com.leetcode.linkedlist;

//https://leetcode.com/problems/remove-duplicates-from-sorted-list-ii/
public class RemoveDuplicatesFromSortedList {

    public ListNode deleteDuplicates(ListNode head) {
        if (head == null) {
            return null;
        }
        ListNode headNode = new ListNode(Integer.MAX_VALUE);
        ListNode currentTemp = headNode;

        int prevValue = -1;

        while (head != null) {
            if (head.next != null) {
                if (head.val == head.next.val || prevValue == head.val) {
                    // move head by 1 pointer as the node is duplicate
                    prevValue = head.val;
                    ListNode temp = head;
                    head = head.next;
                    temp.next = null;
                } else {
                    // append this node to the result
                    currentTemp.next = head;
                    prevValue = head.val;
                    head = head.next;
                    currentTemp = currentTemp.next;
                }
            } else {
                // last node, check if it is duplicate
                if (prevValue == head.val) {
                    currentTemp.next = null;
                } else {
                    currentTemp.next = head;
                }
                head = head.next;
            }
        }
        return headNode.next;
    }

    public static void main(String[] args) {
        ListNode head = new ListNode(1);
        head.next = new ListNode(1);
        head.next.next = new ListNode(1);
        head.next.next.next = new ListNode(2);
        head.next.next.next.next = new ListNode(3);
        //head.next.next.next.next.next = new ListNode(4);
        //head.next.next.next.next.next.next = new ListNode(5);
        ListNode result = new RemoveDuplicatesFromSortedList().deleteDuplicates(head);
        while (result != null) {
            System.out.println(result.val);
            result = result.next;
        }
    }
}
